sdk/communication/azure-communication-callautomation/azure/communication/callautomation/_shared/entra_token_guard_policy.py

```diff
@@ -0,0 +1,88 @@
+# license information.
+# --------------------------------------------------------------------------
+
+import json
+from datetime import datetime, timezone
+from azure.core.pipeline.policies import HTTPPolicy, AsyncHTTPPolicy
+from azure.core.pipeline import PipelineRequest
+
+from dateutil import parser as dateutil_parser
+
+class EntraTokenGuardPolicy(HTTPPolicy):
+    """A pipeline policy that caches the response for a given Entra token and reuses it if valid."""
+
+    def __init__(self):
+        super().__init__()
+        self._entra_token_cache = None
+        self._response_cache = None
+
+    def send(self, request: PipelineRequest):
+        cache_valid, token = _EntraTokenGuardUtils.is_entra_token_cache_valid(self._entra_token_cache, request)
+        if cache_valid and _EntraTokenGuardUtils.is_acs_token_cache_valid(self._response_cache):
+            response = self._response_cache
+        else:
+            self._entra_token_cache = token
+            response = self.next.send(request)
+            self._response_cache = response
+        if response is None:
+            raise RuntimeError("Failed to obtain a valid PipelineResponse in AsyncEntraTokenGuardPolicy.send")
+        return response
+
+
+
+class AsyncEntraTokenGuardPolicy(AsyncHTTPPolicy):
+    """Async pipeline policy that caches the response for a given Entra token and reuses it if valid."""
+
+    def __init__(self):
+        super().__init__()
+        self._entra_token_cache = None
+        self._response_cache = None
+
+    async def send(self, request: PipelineRequest):
+        cache_valid, token = _EntraTokenGuardUtils.is_entra_token_cache_valid(self._entra_token_cache, request)
+        if cache_valid and _EntraTokenGuardUtils.is_acs_token_cache_valid(self._response_cache):
+            response = self._response_cache
+        else:
+            self._entra_token_cache = token
+            response = await self.next.send(request)
+            self._response_cache = response
+        if response is None:
+            raise RuntimeError("Failed to obtain a valid PipelineResponse in AsyncEntraTokenGuardPolicy.send")
+        return response
+    
+    
+
+class _EntraTokenGuardUtils:
+        @staticmethod
+        def is_entra_token_cache_valid(entra_token_cache, request):
+            current_entra_token = request.http_request.headers.get("Authorization", "")
+            cache_valid = (
+                entra_token_cache is not None and
+                current_entra_token == entra_token_cache
+            )
+            return cache_valid, current_entra_token
+
+        @staticmethod
+        def is_acs_token_cache_valid(response_cache):
+            if response_cache is None or response_cache.http_response.status_code != 200:
+                return False
+            return _EntraTokenGuardUtils.is_access_token_valid(response_cache)
+
+        @staticmethod
+        def is_access_token_valid(response_cache):
```

Comment:
nit: The way this is split isn't super intuitive because the abstraction layers (validating on response_cache) are the same.

We can change this method to take the http_response directly or inline it inside its single calling method. Or we can also extract the sub method in a way that it only takes the expires_on after parsing the content.